### ISSUE TITLE
Test tooling enhancements for Nushell scripts

### ISSUE THREAD

**Issue by vyadh** [nushell/nushell#14227]
### Related problem

I think that Nushell has huge potential as a testing tool given it's data manipulation and integration capabilities. Both testing external systems and data as well as for Nushell scripts. Just one example, if you've ever used Nushell with WireMock, using records and `merge` to mock the expected http requests and responses, varying data for different use cases, is just beautiful. I'm very impressed with Nushell's practical focus and how the functionally fits together so well.

Given the importance of testing, it would be great to see the out of the box experience improved with a runner, reporting and test coverage. One reason I'd reach for Nushell over Python for example is, other than being a great integration tool, not having to deal with virtual environments and external dependencies for the 80% case makes it really easy to pick up and run with.

### Describe the solution you'd like

Here's what is ideally like to see: 
- A basic out the box test runner. Something like [testing.nu](https://github.com/nushell/nushell/blob/main/crates/nu-std/testing.nu) used to test the standard library but with output as a table that can then be reported in more flexible ways.
- A default reporter into some common format like Cobertura that can be used in CI/CD systems or sucked into some other converter tool.
- Line-based test coverage metrics.

I can implement the first two points. I'd probably start with the runner for the standard library, put some tests around it, and add support for scripts that aren't modules, all in the hopes that it could be promoted to the Nushell standard library. However, I wonder if that's a bit short sighted. The bit I'd have trouble implementing is the line+file based test coverage, which would likely need to be implemented in the core (or bundled plugin?) and likely needs a far bit of knowledge of Nushell internals.

### Describe alternatives you've considered

Given the ideal scenario of having coverage information, it seems that the strategic option would be to implement some kind of introspection/debugging functionality into the core that can be used to understand the coverage, and a runner and reporter as a plugin or nushell script. Or perhaps just a debugging (coverage-gathering) run mode in the core perhaps behind a `nu` command switch.

A more tactical solution could be done in Nushell without coverage, and perhaps later promoted to core functionality where code coverage could then be added. If we do that right, we could minimise changes to test files, e.g. a `use std testing` statement becomes optional.

Perhaps the core team can provide some guidance / discuss what they would like to do before I start down the tactical road?

### Additional context and details

Prior art:
- [The docs I added](https://www.nushell.sh/book/testing.html#basic-test-framework) for writing a simple test runner.
- The test runner in nupm [here](https://github.com/nushell/nupm/blob/main/nupm/test.nu).
- The [testing.nu](https://github.com/nushell/nushell/blob/main/crates/nu-std/testing.nu) test runner used to run tests for the standard library.
- A test runner I wrote to run the container tests [here](https://github.com/nushell/nushell/blob/main/docker/test_docker.nu), as none of the above was practical to use in that context. You can see the report output for that in PR #14225 .

Perhaps improvements in this area can also help encourage testing nushell scripts in the core. Here are some related issues:
- #5720
- #8696

**Comment by fdncred** [comment]
In general, we want to support testing better. We've talked a long time about using annotations for testing and that testing.nu script supports it, but it's been brittle in the past. Our main goal is to have a lossless parser that can find annotations and then we'd be able to use that information for testing.

In the meantime, we're open for more debug commands that help. Have you seen these?
```nushell
help commands | where category == debug
```

`debug profile` may have some of the items you've mentioned and may be worth a look.

**Comment by vyadh** [comment]
Thanks for the reply and the pointers. That functionality looks far more capable than I had expected. I might throw together a PoC for my own benefit to see whether this could fit a strategic code coverage solution. If that works, do you think having a module in the standard library to do testing and coverage is an acceptable approach? Or do you think it's more likely you'll want something provided by the core? I guess the main benefit of a core-based approach would allow avoiding the need to invoke separate instances of `nu` to run the tests. Perhaps that's not important.

As to annotations being brittle. Do you mean the existing `#[test]` annotations are not always found? Presumably meaning, if any lossless parser that is developed, the way to express annotations won't change, they'll just be more reliable in some way?

**Comment by fdncred** [comment]
> do you think having a module in the standard library to do testing and coverage is an acceptable approach? 

I could live with an internal rust command or a script command. I tend to think rust may be better.

> Do you mean the existing #[test] annotations are not always found

not found, breaks, or just doesn't work for some reason

**Comment by vyadh** [comment]
I've had a play and I think `debug profile` does indeed get us 95% of the way to create a code coverage tool. I imagine there are some edge cases we'd need to parse the IR to figure out (e.g. multi-condition if statements), which seems a bit OTT and brittle but at least possible. I suppose ideally we'd be able to hook into the execution, which again would only be possible from Rust - unless you fancy adding `debug execution-hook` 😉.

Syntax - yeah, definitely agree on just using Rust testing syntax and idioms.

I agree with it ideally being in Rust. I'll have a think about whether I can take that on myself though. It would certainly be more motivating to work a strategic solution than a tactical one, but would be far more challenging one for me personally as I don't know Rust very well.

**Comment by fdncred** [comment]
@kubouch wrote `debug profile`, so, as he has time, he may be able to answer technical questions if you get stuck somewhere but no promises since he's pretty busy. Kubouch has envisioned other commands similar to debug profile as well like adding a debug adapter protocol and things like that.

**Comment by vyadh** [comment]
Thanks. Okay, so adding the basic test runner/reporter functionality as Nushell code and then delving into Rust to add any supporting debug functionality for coverage seems much more achievable for me (over doing everything in Rust). If you're happy with that in principal then I'll start putting something together and you can tell me what you think.

**Comment by kubouch** [comment]
Yes, code coverage should be possible with the `Debugger` trait that is used to implement `debug profile`. It might need some API tweaks, but in principle should be possible. One idea for the trait is to make it possible to connect it to the plugin API so that people would be able to write their own debugging tools as plugins. E.g., a `PluginDebugger` that implements the `Debugger` trait and talks.

The nupm test runner requires running a separate Nushell process and relies on a naming convention. The `testing.nu` requires manual Nushell parsing of the comments and is very slow and error-prone. Neither solution is ideal, we probably need to extend the language to make testing a first-class citizen.

**Comment by vyadh** [comment]
Thanks for that. I previously looked at `testing.nu` and the main issue seems to be that it's invoking nu for every test. I noticed a massive speed up by first creating a test plan and invoking nu just once for all tests. You can where I've implemented that technique in [the Docker tests here](https://github.com/nushell/nushell/blob/main/docker/test_docker.nu#L17) but that has only a few tests. I've done it elsewhere with a far larger number of tests and it indeed makes a huge difference.

I think I can best contribute creating the functionality of a general purpose test runner and reporter with test coverage as a Nushell module (raising any limitations due to missing language features as I go). However, if "first class" means implementing this in Rust then I think that's probably a bit beyond the time investment I can provide atm given I'd need to learn Rust and the Nushell internals as well. If "first class" can be a Nushell module in the standard library, which highlights any limitations and supporting rust functionality is added as deemed suitable, then I'm definitely down for doing that. In fact, I've started it, but would rather focus on other areas of Nushell if this is deemed better in the Rust core.

**Comment by vyadh** [comment]
Is there a `complete` for internal commands? I would like to be able to report any output or error messages from a test, but there doesn't seem to be a function for this.

For example, I'd like to be able to do something like:

```nu
def some-test [] {
  print "some output"
}

let $result = do-capture { some-test }
```

Where `$result` might then be something like:

```nu
{
  result: <same as normal do command>
  output: ...
  error: ...
}
```

Does something like this exist?

One workaround would be to shell out to  `nu -c` to capture the test output but that seems tricky in my current design as I'm trying to see if I can run the tests on a per-"test suite" basis rather than a per-test one. Shelling out to nu for each test would lose the test context, and perhaps cause problems given I'd likely need to separate out unstructured output from nuon-serialised test results. And of course, far worse performance when you have lots of tests such as a project using a TDD-based approach.

**Comment by fdncred** [comment]
You might experiment with `try/catch`. I recently made some changes to it where one can manipulate errors better https://github.com/nushell/nushell/pull/14082. I'm not sure it'll work for what you want, but it's worth a look.

**Comment by vyadh** [comment]
I saw you added that. That's super helpful to avoid parsing the error debug message as I'm currently doing.

The main issue I have with output via `print` is that it breaks the nuon test results passed back to the parent `nu` process. Still, this doesn't seem to be a new issue as the existing `testing.nu` also discards any test output. I think perhaps in an initial implementation I can just document that any output / debugging should be done via `print -e` and I'll come back it to later.

**Comment by fdncred** [comment]
Maybe `print` isn't the right thing to use since it just prints "dumb strings" and not nushell values? You can return values with `$value` or `echo $value`, assuming value is your variable name, but you can only return one thing per custom command.

**Comment by vyadh** [comment]
Agreed. I meant any use of `print` by people writing tests given it interferes with what is returned from the nu subprocess. I'll just document that doing this isn't supported until Nushell has some kind of output capture/suppression of internal commands, assuming we want that.

**Comment by weirdan** [comment]
What if you overrode `print` in the subprocess?

**Comment by vyadh** [comment]
I totally didn't think of that. That works perfectly. Thank you!

**Comment by vyadh** [comment]
Just an update, I've made fairly good progress with the initial version of this: https://github.com/vyadh/nu-test
It still needs a few tweaks and clean-ups, some more testing as well as some updates to integrate it into the Nushell standard library build before I submit a PR and we can discuss. Then better docs because the readme is a bit of a mess (being more like my todo list and thought process). Test coverage (as a guess) is probably more than 95%.

I guess this will be on the larger side as standard library module, if accepted, but the actual functionality is < 600 lines of Nushell code. Not too much bigger than `testing.nu`.

There are a few bigger things missing - like adding a standard reporter format (e.g. TRX, JUnit or nextest) and a way of measuring coverage. I'll do that once we're happy with the basic functionality.

**Comment by vyadh** [comment]
I've completed the first version of [this Nushell testing module](https://github.com/vyadh/nu-test) (it took a while to nail various edge cases!). Should I create a PR on the main Nushell repo or in the new standard library candidate space?

I did have trouble trying to create a working nu binary because of the multiple files that make up this module, but I'll work through that when I know where it's going.

**Comment by vyadh** [comment]
The PR  https://github.com/nushell/nu_scripts/pull/1002 adds testing.nu to allow testing of standard lib candidates so now seems like a good time to ask whether a test framework should be added to the standard library / candidates or not.

A basic one such as testing.nu would be better than nothing. The main advantage being a small amount of code. The main disadvantage being lack of output on what failed and it doesn't have tests for itself. 

Another is [nu-test](https://github.com/vyadh/nu-test). The main disadvantage of that I'd say is that it's now a lot more code.

I suppose another question is whether we would add this if we had an effective and out the box package manager.

**Comment by vyadh** [comment]
It would be neat to have an out of the box test runner. I thought that would be [Nutest](https://github.com/vyadh/nutest). However, I think adding it to the stdlib with it's comprehensive test suite might mean that valid behaviour changes in Nushell might too often require adjusting Nutest tests just to keep Nutest happy, which I'm sure would be annoying for contributors.

I've released nutest v1.0 now and barring the current lack of test coverage it's fully functional and can output CI reports as well as tables for querying, etc, so I'm going to close this issue because Nushell now at least has a flexible test runner, which I'm happy to enhance and maintain against future Nushell releases. It could still be used to run tests for the standard library in Nushell's CI if that was deemed useful - in fact it already does this in the Nutest CI tests so that I'm alerted to any issues in Nutest from Nushell `nightly` that need to be fixed before the next Nushell release.

I'm still happy to add this as a standard library candidate if there is demand. Perhaps I could strip the tests down to a set of functional integration tests to avoid the above issue for example. However, I do wonder whether if we had a really good story around nupm there would be less pressure to add things to the stdlib, or indeed Nushell itself. We can keep things in the "core" leaner because other stuff like nutest is just a `nupm install` away. Nushell is batteries included, but maybe it's as AA's than DD's 😄
